### DISCUSSION TITLE
Integration Testing of Rust/QML/Javascript code

### DISCUSSION THREAD

**Discussion by akiselev** [KDAB/cxx-qt#961] [Q&A] [answered]
Is there an example of how to test cxx-qt code that requires a QCoreApplication?

I've wrapped QJSEngine/QJSValue/etc. and I'm trying to write serialization/deserialization tests but I'm having problems with QCoreApplication engine with the following errors (after restricting the number of threads to 1 so its all on the main thread):

```
test deserializer::tests::test_deserialize_bool ...
QCoreApplication::applicationDirPath: Please instantiate the QApplication object first
QEventLoop: Cannot be used without QApplication
QJSEngine: Must construct a QCoreApplication before a QJSEngine

```

This is the test code I'm trying to use:

``` rust
    fn init() -> UniquePtr<QQmlApplicationEngine> {
        QQmlApplicationEngine::new()
    }

    /// Setup function to create a QJSEngine and evaluate a script
    fn setup_and_evaluate(script: &str) -> (UniquePtr<QJSEngine>, UniquePtr<QJSValue>) {
        let mut engine_ptr = QJSEngine::new();
        let engine = engine_ptr.as_mut().unwrap();
        let value = engine.evaluate(&QString::from(script), &QString::from(""), 1);
        (engine_ptr, value)
    }

    #[test]
    fn test_deserialize_bool() {
        let _ = init();
        let (mut engine, value) = setup_and_evaluate("true");
        let deserialized: bool = engine.as_mut().unwrap().deserialize(&value).unwrap();
        assert_eq!(deserialized, true);
    }
```

I believe in C++ there's macros like `QTEST_GUILESS_MAIN` but I'm not sure how to implement that in Rust. Subclass `QTest` and use it as a test runner via `qExec` or something?

**Comment by akiselev** [comment, marked as answer]
I was instantiating `QQmlApplicationEngine` instead of `QCoreApplication` :facepalm:

**Reply by dphaldes** [reply]
Unrelated but @akiselev did you write your own wrappers around QJSValue and QJSEngine?

**Reply by akiselev** [reply]
@mystchonky yep. Are there existing ones by any chance?

**Reply by dphaldes** [reply]
Not ATM from what I can see. I am in the middle of writing my own and maybe upstream it. Can you share yours ?

**Reply by akiselev** [reply]
I just started a few days ago and I'm just adding functionality as I use it in my app so yours is probably further along. I can share the implementation in a gist or something but its not much

**Reply by ahayzen-kdab** [reply]
(Also BTW, feel free to submit types to cxx-qt-lib or cxx-qt-lib-extras, which we have for less supported/API ready types, if they are useful :-) )
